### PR TITLE
LLM Finetune with PEFT

### PR DESCRIPTION
**Pull request by avishniakov**
This PR brings in the pipeline to run Mistral model fine-tuning using PEFT library on Viggio dataset.

Key highlights:
- Project has one pipeline to run the full cycle of LLM fine-tuning:
    - Data extraction and preparation
    - Finetuning base model on new data
    - Evaluation of the performances for base and fine-tuned model
    - Promotion of the model to the target environment, if it outperforms the base and currently promoted one
- Model can be trained on local(remote) orchestrator in full
- Model can be trained on step operators, where only GPU intense steps will be pushed to step operator (we used Vertex AI step operator)
- Here we also move around the previous LitGPT fine-tune project to `llm-litgpt-finetuning` and this new project becomes `llm-lora-finetuning`

P.S. Diff is a bit off due to project movement. LitGPT one was lift and shift - no changes in it.

Template update: https://github.com/zenml-io/template-llm-finetuning/pull/4

### CODE REVIEW

**Comment by strickvl** [comment]
@coderabbitai review

File: llm-peft-finetune.tar.gz
Comment (strickvl) [inline llm-peft-finetune.tar.gz:1]: what's this?

File: llm-lora-finetuning/README.md [hunk 1]
```diff
@@ -15,13 +15,13 @@ Using these pipelines, we can run the data-preparation and model finetuning with
   <br/>
 </div>
 
-## :earth_americas: Inspiration and Credit
+## 🌎 Inspiration and Credit
 
-This project heavily relies on the [Lit-GPT project](https://github.com/Lightning-AI/litgpt) of the amazing people at Lightning AI. We used [this blogpost](https://lightning.ai/pages/community/lora-insights/#toc14) to get started with LoRA and QLoRA and modified the commands they recommend to make them work using ZenML.
+This project heavily relies on the [PEFT project](https://huggingface.co/docs/peft/en/index) of the amazing people at Huggingface and [mistralai/Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1) model from the amazing people at MistralAI.
```
Comment (strickvl) [inline llm-lora-finetuning/README.md:20]:
```suggestion
This project heavily relies on the [PEFT project](https://huggingface.co/docs/peft/en/index) by the amazing people at Huggingface and the [`mistralai/Mistral-7B-v0.1`](https://huggingface.co/mistralai/Mistral-7B-v0.1) model from the amazing people at MistralAI.
```

File: llm-lora-finetuning/materializers/directory_materializer.py [hunk 2]
```diff
@@ -52,6 +53,7 @@ def save(self, data: Any) -> None:
         """
         assert isinstance(data, Path)
         self._copy_directory(src=str(data), dst=self.uri)
+        shutil.rmtree(data)  # clean-up stored locally data
```
Comment (strickvl) [inline llm-lora-finetuning/materializers/directory_materializer.py:56]:
```suggestion
        shutil.rmtree(data)  # clean-up locally stored data
```

File: llm-lora-finetuning/run.py [hunk 3]
```diff
@@ -14,77 +14,36 @@
...
-)
-from zenml.logger import get_logger
-
-logger = get_logger(__name__)
+from pipelines.train import llm_peft_full_finetune
 
 
 @click.command(
     help="""
 ZenML LLM Finetuning project CLI v0.1.0.
```
Comment (strickvl) [inline llm-lora-finetuning/run.py:26]:
```suggestion
ZenML LLM Finetuning project CLI v0.2.0.
```

Probably? :)

File: llm-lora-finetuning/steps/evaluate_model.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+)
+from utils.tokenizer import tokenize_for_eval
+from zenml import log_model_metadata, save_artifact, step
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
+
+
+@step
+def evaluate_model(
```
Comment (strickvl) [inline llm-lora-finetuning/steps/evaluate_model.py:37]: maybe a docstring for this and other steps explaining what they do?

File: llm-lora-finetuning/utils/loaders.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
+from transformers import AutoModelForCausalLM, BitsAndBytesConfig
+from utils.logging import print_trainable_parameters
+from utils.tokenizer import generate_and_tokenize_prompt, load_tokenizer
+
+
+def prepare_data(
+    base_model_id: str,
+    system_prompt: str,
+    is_tain: bool = True,
```
Comment (strickvl) [inline llm-lora-finetuning/utils/loaders.py:33]: is_train (and fixes elsewhere as a result?)

**Review by strickvl — CHANGES_REQUESTED** [review]
Initial quick read. Looks good. Will read more thoroughly on Monday. Nice work on this!

File: llm-peft-finetune.tar.gz
Comment (avishniakov) [reply llm-peft-finetune.tar.gz:1]: Left over, removed

File: llm-lora-finetuning/utils/loaders.py (hunk 5 above)
Comment (avishniakov) [reply llm-lora-finetuning/utils/loaders.py]: This method was not used - I used step to do it, so just removed the whole thing

**Review by strickvl — COMMENTED** [review]
Not much else to say beyond these comments (and the ones from before). Looks good overall.

File: llm-lora-finetuning/requirements.txt [hunk 6]
```diff
@@ -1,17 +1,10 @@
...
-# scripts/prepare_slimpajama.py, scripts/prepare_starcoder.py, pretrain/tinyllama.py
-lightning[data] @ git+https://github.com/Lightning-AI/lightning@ed367ca675861cdf40dbad2e4d66f7eee2ec50af
+datasets
+transformers
+peft
+bitsandbytes==0.41.0
+scipy
+evaluate
+rouge_score
+nltk
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply llm-lora-finetuning/requirements.txt:10]: Not directly, but it is needed by ROUGE and it is poorly done for deps installation, so have to be explicit here.

File: llm-lora-finetuning/utils/callbacks.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            return
+
+    def on_save(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        # TODO: add ability to save model checkpoints here, will likely get redundant with Mounts
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply llm-lora-finetuning/utils/callbacks.py:56]: Yeah, not in current iteration, but worth doing a bit later, IMO

File: llm-lora-finetuning/steps/evaluate_model.py (hunk 4 above)
Comment (avishniakov) [reply llm-lora-finetuning/steps/evaluate_model.py:37]: Added

**Review by strickvl — APPROVED** [review]

File: llm-lora-finetuning/utils/logging.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
+
+
+def print_trainable_parameters(model):
```
Comment (schustmi) [inline llm-lora-finetuning/utils/logging.py:23]: Missing type annotation for model/return type

File: llm-lora-finetuning/pipelines/train.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    
+    It will run the following steps:
+
+    - configure: set the system prompt and base model id
+    - prepare_data: prepare the datasets and tokenize them
+    - finetune: finetune the model
+    - evaluate_model: evaluate the base and finetuned model
+    - promote: promote the model to the target stage, if evaluation was successful
+    """
+    system_prompt, base_model_id = configure()
```
Comment (schustmi) [inline llm-lora-finetuning/pipelines/train.py:48]: I'm wondering, why have these as a separate step and not as pipeline parameters? So we store them in the artifact store for visibility?

File: llm-lora-finetuning/pipelines/train.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+zenml_logging.STEP_LOGS_STORAGE_MAX_MESSAGES = (
+    10000  # workaround for https://github.com/zenml-io/zenml/issues/2252
+)
+
+
+@pipeline(
+    settings={
+        "docker": DockerSettings(
+            parent_image="pytorch/pytorch:2.2.2-cuda11.8-cudnn8-runtime",
```
Comment (schustmi) [inline llm-lora-finetuning/pipelines/train.py:32]: I had to switch the image when switching from AWS to GCP as they require different CUDA versions in their VM. Maybe it makes sense to have that in the `config.yaml` as well for people to easily modify according to which version they need?

**Review by schustmi — CHANGES_REQUESTED** [review]
Just a few tiny nits, lgtm otherwise

File: llm-lora-finetuning/pipelines/train.py (hunk 10 above)
Comment (avishniakov) [reply llm-lora-finetuning/pipelines/train.py]: Done

File: llm-lora-finetuning/pipelines/train.py (hunk 9 above)
Comment (avishniakov) [reply llm-lora-finetuning/pipelines/train.py]: Moved it to logging in a form of metadata

File: llm-lora-finetuning/utils/logging.py (hunk 8 above)
Comment (avishniakov) [reply llm-lora-finetuning/utils/logging.py]: added

**Review by schustmi — APPROVED** [review]

**Comment by avishniakov** [comment]
@htahir1 are we ok to merge this? It will replace the current LORA example with PEFT one and LitGPT will be moved to a new directory. Any blockers?

**Comment by schustmi** [comment]
Just one more thing that came to my mind: We should probably also rename the `llm-lora-finetuning` template right?

**Review by htahir1 — APPROVED** [review]

**Pull request merged**